v2/services/authorize/secret.go

```diff
@@ -2,30 +2,157 @@ package authorize
 
 import (
 	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
 	"fmt"
+	"net"
+	"time"
 
-	googleGrpc "google.golang.org/grpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
 
-	"github.com/SKF/go-enlight-sdk/v2/grpc"
 	"github.com/SKF/go-enlight-sdk/v2/services/authorize/credentialsmanager"
 )
 
-func getCredentialOption(ctx context.Context, host, secretKeyName string, cm *credentialsmanager.CredentialsManager) (googleGrpc.DialOption, error) {
-	clientCert, err := cm.GetDataStore(ctx, secretKeyName)
+func getCredentialOption(ctx context.Context, host, secretKeyName string, cf credentialsmanager.CredentialsFetcher) (grpc.DialOption, error) {
+	c, err := NewAutoRefreshingTransportCredentials(ctx, cf, secretKeyName, host)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
-	return grpc.WithTransportCredentialsPEM(
-		host,
-		clientCert.Crt, clientCert.Key, clientCert.CA,
-	)
+	return grpc.WithTransportCredentials(c), nil
 }
 
-func GetSecretKeyName(service, stage string) string {
-	return fmt.Sprintf("authorize/%s/grpc/client/%s", stage, service)
+type autoRefreshingTransportCredentials struct {
+	credentials           credentials.TransportCredentials
+	cf                    credentialsmanager.CredentialsFetcher
+	secretKeyName         string
+	serverName            string
+	certificateExpiryTime time.Time
 }
 
-func GetSecretKeyArn(accountId, region, service, stage string) string {
-	return fmt.Sprintf("arn:aws:secretsmanager:%s:%s:secret:%s", region, accountId, GetSecretKeyName(service, stage))
+func NewAutoRefreshingTransportCredentials(ctx context.Context, cf credentialsmanager.CredentialsFetcher, secretKeyName, host string) (credentials.TransportCredentials, error) {
+	creds := &autoRefreshingTransportCredentials{
+		cf:            cf,
+		secretKeyName: secretKeyName,
+		serverName:    host,
+	}
+
+	err := creds.loadCertificates(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return creds, nil
+}
+
+func (c *autoRefreshingTransportCredentials) ensureValidCredentials(ctx context.Context) error {
+	if c.shouldLoadNewCertificates() {
+		if err := c.loadCertificates(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
Maybe we could rearrange this to make it more shallow?
```suggestion
	if !c.shouldLoadNewCertificates() {
		return nil
	}
	
    if err := c.loadCertificates(ctx); err != nil {
		return err
	}
	
	return nil
```

Reply:
Sure! Updated.